cpp/src/parquet/arrow/reader.h

```diff
@@ -249,6 +249,90 @@ class PARQUET_EXPORT FileReader {
+  ///
+  /// \see ReadRowGroupsAsync for operation details
+  /// \see ReadTable for details on how column indices are resolved
+  ///
+  /// \param i the index of the row group to read
+  /// \param column_indices leaf-indices of the columns to read
+  /// \param cpu_executor an executor to use to run CPU tasks
+  /// \param allow_sliced_batches if false, an error is raised if a batch has too much
+  ///                             data for the given batch size.  If true, smaller
+  ///                             batches will be returned instead.
+  virtual AsyncBatchGenerator ReadRowGroupAsync(int i,
+                                                const std::vector<int>& column_indices,
+                                                ::arrow::internal::Executor* cpu_executor,
+                                                bool allow_sliced_batches = false) = 0;
+
+  /// \brief Read row groups from the file
+  ///
+  /// \see ReadRowGroupsAsync for operation details
+  ///
+  /// \param row_groups indices of the row groups to read
+  /// \param cpu_executor an executor to use to run CPU tasks
+  /// \param allow_sliced_batches if false, an error is raised if a batch has too much
+  ///                             data for the given batch size.  If true, smaller
+  ///                             batches will be returned instead.
+  virtual AsyncBatchGenerator ReadRowGroupsAsync(
+      const std::vector<int>& row_groups, ::arrow::internal::Executor* cpu_executor,
+      bool allow_sliced_batches = false) = 0;
+
+  /// \brief Read some columns from the given rows groups from the file
+  ///
+  /// If pre-buffering is enabled then all of the data will be read using the pre-buffer
+  /// cache. See ParquetFileReader::PreBuffer for details on how this affects memory and
+  /// performance.
+  ///
+  /// This operation is not perfectly async.  The read from disk will be done on an I/O
+  /// thread, which is correct.  However, compression and column decoding is also done on
+  /// the I/O thread which may not be ideal.  The stage after that (transferring the
+  /// decoded data into Arrow structures and fulfilling the future) should be done as a
+  /// new task on the cpu_executor.
+  ///
+  /// The returned generator will respect the batch size set in the ArrowReaderProperties.
+  /// Batches will not be larger than the given batch size.  However, batches may be
+  /// smaller.  This can happen, for example, when there is not enough data or when a
+  /// string column is too large to fit into a single batch.  The parameter
+  /// `allow_sliced_batches` can be set to false to disallow this later case.  This can be
+  /// useful when you need to know exactly how many batches you will get from the
+  /// operation before you start.
+  ///
+  /// Note: When reading multiple row groups there is no guarantee you will get one
+  /// record batch per row group.  Data from multiple row groups could get combined into
+  /// a single batch.
```

Comment:
Interesting, and I agree it's probably desirable. Is it a deviation from other APIs?

Reply:
No.  This is the same way that the synchronous APIs behave.